crates/iota-data-ingestion-core/src/reader/v2.rs

```diff
@@ -0,0 +1,717 @@
+// Copyright (c) 2025 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    future::Future, num::NonZeroUsize, ops::Range, path::PathBuf, sync::Arc, time::Duration,
+};
+
+use backoff::backoff::Backoff;
+use futures::{StreamExt, future::OptionFuture};
+use iota_config::{
+    node::ArchiveReaderConfig,
+    object_storage_config::{ObjectStoreConfig, ObjectStoreType},
+};
+use iota_metrics::spawn_monitored_task;
+use iota_rest_api::CheckpointData;
+use iota_storage::blob::Blob;
+use iota_types::messages_checkpoint::CheckpointSequenceNumber;
+use object_store::{ObjectStore, path::Path as ObjectStorePath};
+use tokio::{
+    sync::{
+        mpsc::{self, error::TryRecvError},
+        oneshot,
+    },
+    task::JoinHandle,
+    time::timeout,
+};
+use tracing::{debug, error, info};
+
+use crate::{
+    IngestionError, IngestionResult, MAX_CHECKPOINTS_IN_PROGRESS, create_remote_store_client,
+    history::reader::HistoricalReader,
+    reader::{
+        fetch::{
+            CheckpointResult, ReadSource, fetch_from_full_node, fetch_from_object_store,
+            read_local_files_with_retry_and_capacity_check, setup_directory_watcher,
+        },
+        v1::{DataLimiter, gc_processed_files},
+    },
+};
+
+/// Available sources for checkpoint streams supported by the ingestion
+/// framework.
+///
+/// This enum represents the different types of remote sources from which
+/// checkpoint data can be fetched. Each variant corresponds to a supported
+/// backend or combination of backends for checkpoint retrieval.
+pub enum RemoteUrl {
+    /// A REST API endpoint for checkpoint data.
+    Rest(String),
+    /// An object storage backend for checkpoint data.
+    ObjectStore {
+        /// The URL of the object store (e.g., S3 bucket, GCS bucket,
+        /// HTTP/WebDAV Storage).
+        object_store_url: String,
+        /// Additional options for configuring the remote store as key-value
+        /// pairs.
+        remote_store_options: Vec<(String, String)>,
+    },
```

Comment:
This is the one that we want to deprecate, no? So why to include it in the new reader?